python/diff_frames.py

```diff
@@ -96,62 +259,158 @@ def _do_diff(self):
-        diff_line_width = self._terminal.width - duration_width - 1
+        self._draw_state = ds = _DrawState(self)
+        ds.end_times = self._durations.map(str)
+        diff_lines = self._make_diff_lines(ds, self._diff)
+        if self._zoom == 1:
+            pass
+        else:
+            # FIXME: Show times of each end when zoomed in
+            pass
         lines = [
-            self._make_diff_line(
-                self._diff, self._len, diff_line_width) + ' ' + duration_a,
-            self._make_diff_line(
-                self._diff, self._len, diff_line_width, is_b=True) + ' ' +
-            duration_b
+            diff_lines.a,
+            self._make_cue_line(ds),
+            diff_lines.b,
+            self._status.ljust(self._terminal.width)
         ]
         self._terminal.print_lines(lines)
 
+    def _zoom_in(self):
+        self._zoom *= 1.1
+
+    def _zoom_out(self):
+        self._zoom /= 1.1
+
+    def _cue_hunk_helper(self, iterable, predicate):
+        self._active_cue = None
+        for hunk in iterable:
+            if predicate(hunk):
+                with self._free_cues():
+                    self._start_cue = hunk.start
+                    self._end_cue = hunk.end
+                self._cursor = self._start_cue
+                break
+
+    def _cue_next_hunk(self):
+        self._cue_hunk_helper(
+            self._diff, lambda hunk: hunk.start > self._cursor)
+
+    def _cue_prev_hunk(self):
+        self._cue_hunk_helper(
+            reversed(self._diff), lambda hunk: hunk.end < self._cursor)
+
+    def _select_cue_helper(self, cue_name):
+        if self._active_cue == cue_name:
+            self._active_cue = None
+        else:
+            self._active_cue = cue_name
+
+    def _select_start_cue(self):
+        self._select_cue_helper('start')
+
+    def _select_end_cue(self):
+        self._select_cue_helper('end')
+
+    def _move_active_cue_point(self, d):
+        attr_name = '_{}_cue'.format(self._active_cue)
+        setattr(
+            self, attr_name,
+            getattr(self, attr_name) + self._draw_state.to_frames(d))
+
+    def _move_cursor(self, d):
+        self._cursor += self._draw_state.to_frames(d)
+
+    def _on_arrow(self, d):
+        if self._active_cue:
+            self._move_active_cue_point(d)
+        else:
+            self._move_cursor(d)
+
+    def _on_left(self):
+        self._on_arrow(-1)
+
+    def _on_right(self):
+        self._on_arrow(1)
+
+    def _on_home(self):
+        if self._cursor == self._start_cue:
+            self._cursor = 0
+        else:
+            self._cursor = self._start_cue
+
+    def _on_end(self):
+        if self._cursor == self._end_cue:
+            self._cursor = self._len
+        else:
+            self._cursor = self._end_cue
+
+    def _on_esc(self):
+        if self._active_cue:
+            self._active_cue = None
+        else:
+            self._cursor = self._start_cue
```

Comment:
It is kinda idempotent. If you keep mashing escape it gets you back to a point where nothing changes. When you're in 'cue moving' mode (`self._active_cue != None`) wouldn't you expect `esc` to get you out of that mode. Wouln't it then be a pain in normal mode to have a different key binding to move the cursor back to the cue point? I thought `esc esc` for "move cue and jump" was rather intuitive?